provider/pkg/provider/provider_parameterize.go

```diff
@@ -0,0 +1,367 @@
+}
+
+// get retrieves the CRD schema for the given package name.
+func (c *parameterizedPackageMap) get(name, version string) *pulumischema.PackageSpec {
+	c.Lock()
+	defer c.Unlock()
+	return c.crdSchemas[c.mapKey(name, version)]
+}
+
+// add adds the PackageSpec for a given parameterized package.
+func (c *parameterizedPackageMap) add(name, version string, schema *pulumischema.PackageSpec) {
+	c.Lock()
+	defer c.Unlock()
+	if c.crdSchemas == nil {
+		c.crdSchemas = make(map[string]*pulumischema.PackageSpec)
+	}
+
+	c.crdSchemas[c.mapKey(name, version)] = schema
+}
+
+// ParameterizedArgs is the struct that holds the arguments for the Kubernetes Provider parameterization.
+// Currently, this is only used for generating types from CRD manifests.
+type ParameterizedArgs struct {
+	PackageVersion   string
+	CRDManifestPaths []string
+}
+
+// String returns a string representation of the ParameterizedArgs. This is used for logging purposes.
+func (p ParameterizedArgs) String() string {
+	return fmt.Sprintf("version: %s, crd-manifests: %v", p.PackageVersion, strings.Join(p.CRDManifestPaths, ", "))
+}
+
+// parseCrdArgs parses the user provided arguments for provider parameterization.
+func parseCrdArgs(args []string) (*ParameterizedArgs, error) {
+	var crdPackageVersion string
+	var yamlPaths []string
+
+	flags := pflag.NewFlagSet("crdargs", pflag.PanicOnError)
+	flags.StringVarP(&crdPackageVersion, "version", "v", "", "The version of the CRD package.")
+	flags.StringArrayVarP(&yamlPaths, "crd-manifests", "c", nil, "The paths to the CRD manifests.")
+	err := flags.Parse(args)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing parameterization args: %w", err)
+	}
+
+	if crdPackageVersion == "" {
+		return nil, errors.New("package version must be provided")
+	}
+
+	if len(yamlPaths) == 0 {
+		return nil, errors.New("no locations of yaml files given")
+	}
+
+	return &ParameterizedArgs{
+		PackageVersion:   crdPackageVersion,
+		CRDManifestPaths: yamlPaths,
+	}, nil
+}
+
+// fillDefaultNames sets the default names for the CRD if they are not specified.
+// This allows the OpenAPI builder to generate the swagger specs correctly with
+// the correct defaults.
+func setCRDDefaults(crd *extensionv1.CustomResourceDefinition) {
+	if crd.Spec.Names.Singular == "" {
+		crd.Spec.Names.Singular = strings.ToLower(crd.Spec.Names.Kind)
+	}
+	if crd.Spec.Names.ListKind == "" {
+		crd.Spec.Names.ListKind = crd.Spec.Names.Kind + "List"
+	}
+}
+
+// crdToOpenAPI generates the OpenAPI specs for a given CRD manifest.
+func crdToOpenAPI(crd *extensionv1.CustomResourceDefinition) ([]*spec.Swagger, error) {
+	var openAPIManifests []*spec.Swagger
+
+	setCRDDefaults(crd)
+
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			continue
+		}
+		// Defaults are not pruned here, but before being served.
+		sw, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true, StripValueValidation: false, StripNullable: false, AllowNonStructural: true})
```

Comment:
Schemagen currently parses an OpenAPI v2 swagger spec (https://github.com/pulumi/pulumi-kubernetes/blob/f5f368dad365317f265f484b2d61a9180ee352f6/Makefile#L15). I didn't look too deep, but refactoring schemagen to handle v3 will likely be a larger effort. Since there are no plans for K8s to deprecate serving swagger specs, I'm deferring that work for now.